portal-impl/src/com/liferay/portal/security/auth/AuthVerifierPipeline.java

```diff
@@ -102,169 +133,215 @@ private AuthVerifierResult _createGuestVerificationResult(
+				filterProperties.entrySet()) {
 
-			return false;
-		}
+			String propertyName = propertyEntry.getKey();
+			Object propertyValue = propertyEntry.getValue();
 
-		String[] urlsIncludes = StringUtil.split(
-			properties.getProperty("urls.includes"));
+			if (propertyName.startsWith(authVerifierPropertyName) &&
+				(propertyValue instanceof String)) {
 
-		if (urlsIncludes.length == 0) {
-			return false;
+				mergedProperties.setProperty(
+					propertyName.substring(authVerifierPropertyName.length()),
+					(String)propertyValue);
+			}
 		}
 
-		if (Wildcard.matchOne(requestURI, urlsIncludes) > -1) {
-			return true;
+		if (mergedProperties.size() < 1) {
+			if (filterProperties.containsKey("portal_property_prefix")) {
+				return authVerifierConfiguration;
+			}
+
+			return null;
 		}
 
-		return false;
+		AuthVerifierConfiguration mergedAuthVerifierConfiguration =
+			new AuthVerifierConfiguration();
+
+		mergedAuthVerifierConfiguration.setAuthVerifier(
+			authVerifierConfiguration.getAuthVerifier());
+		mergedAuthVerifierConfiguration.setAuthVerifierClassName(
+			authVerifierConfiguration.getAuthVerifierClassName());
+		mergedAuthVerifierConfiguration.setProperties(mergedProperties);
+
+		return mergedAuthVerifierConfiguration;
 	}
 
-	private AuthVerifierConfiguration _mergeAuthVerifierConfiguration(
-		AuthVerifierConfiguration authVerifierConfiguration,
-		AccessControlContext accessControlContext, String contextPath) {
+	private void _rebuildConfiguration() {
+		Map<String, List<AuthVerifierConfiguration>>
+			excludeAuthVerifierConfigurations = new HashMap<>();
 
-		Map<String, Object> settings = accessControlContext.getSettings();
+		Map<String, List<AuthVerifierConfiguration>>
+			includeAuthVerifierConfigurations = new HashMap<>();
 
-		String authVerifierSettingsKey = getAuthVerifierPropertyName(
-			authVerifierConfiguration.getAuthVerifierClassName());
+		for (AuthVerifierConfiguration authVerifierConfiguration :
+				AuthVerifierTrackerCustomizer._authVerifierConfigurations) {
 
-		boolean merge = false;
+			authVerifierConfiguration = _mergeAuthVerifierConfiguration(
+				authVerifierConfiguration, _filterProperties);
 
-		Set<String> settingsKeys = settings.keySet();
+			if (authVerifierConfiguration == null) {
+				continue;
+			}
 
-		Iterator<String> iterator = settingsKeys.iterator();
+			Properties properties = authVerifierConfiguration.getProperties();
 
-		while (iterator.hasNext() && !merge) {
-			String settingsKey = iterator.next();
+			String[] urlsExcludes = StringUtil.split(
+				properties.getProperty("urls.excludes"));
 
-			if (settingsKey.startsWith(authVerifierSettingsKey) &&
-				(settings.get(settingsKey) instanceof String)) {
+			for (String urlsExclude : urlsExcludes) {
+				urlsExclude = _fixLegacyURLPattern(urlsExclude);
 
-				merge = true;
-			}
-		}
+				excludeAuthVerifierConfigurations.computeIfAbsent(
+					urlsExclude, key -> new ArrayList<>());
```

Comment:
@arthurchan35, 

I was thinking if it would make sense to first process `url.includes` and then process `url.excludes` if and only if the url in `url.excludes` is included in at least one of the `url.includes`. I believe if they don't we should not even track it since they would never be applied, would they?

Reply:
I think I am bit against the idea, because doing so would imply that we are doing verification for devs' code, I think it's devs responsibility for making sure url.excludes are corresponding to url.includes. And I think performance gain is probably mimial.

Also if we want to do it, instead of testing if the **url in url.excludes is included in at least one of the url.includes**, it should be:
is any url in url.includes, after removing the sufix of wildcard, the start of the given url in url.excludes.

Reply:
ok... I don't have a strong opinion on this. My intention was not to test devs code, I just thought we could use the `include.urls` `URLMapper` to check if the `exclude.urls` match. It looked like a harmless and easy check to me and it could also maybe allow us to check first for `include.urls` in the request, so for most of requests we can save one search. 

But `exclude.urls` are gonna be empty most of the times anyway.